### PR TITLE
String operators

### PR DESCRIPTION
**Pull request by RedBrumbler**
Adds various operators to StringW to make it faster and abstract checks behind operators

 - operator==
 - operator<
 - operator+
 - operator+=
 - operator[]
 - starts_with
 - ends_with
 
 It's probably best to merge PR #101 before merging this one

### CODE REVIEW

File: shared/utils/typedefs-string.hpp [hunk 1]
```diff
@@ -83,13 +86,25 @@ struct ConstString {
...
 
+    friend StringW;
     private:
     void* klass = nullptr;
     void* monitor;
     int length;
     char16_t chars[sz + 1];
 };
 
+bool streq(Il2CppString const* lhs, const std::string_view rhs) noexcept;
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Can you move this (and all the other function declarations) to the `il2cpp_utils::detail` namespace, see elsewhere in this file.

File: shared/utils/typedefs-string.hpp [hunk 2]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
     constexpr Il2CppString* operator->() noexcept {
         return inst;
     }
-    operator std::u16string();
-    operator std::wstring();
+    constexpr operator bool() const noexcept {
+        return inst != nullptr; 
+    }
+
+    // the argument here is unused but we don't care. the comparison against nullptr more clearly shows what this operator is for imo
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: try using `[[maybe_unused]]` or just not giving the parameter a name, though you should actually compare it to the parameter to be clear.

File: shared/utils/typedefs-string.hpp [hunk 3]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wunused-parameter"
+    constexpr bool operator ==(std::nullptr_t rhs) const noexcept {
+        return inst == nullptr;
+    }
+    #pragma clang diagnostic pop
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Might make more sense for this constraint to essentially say: "can I construct a StringW from this" instead

File: shared/utils/typedefs-string.hpp [hunk 4]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+    StringW& operator +=(const T rhs) noexcept {
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: East const ;)

File: shared/utils/typedefs-string.hpp (hunk 1 above)
Comment (sc2ad) [reply shared/utils/typedefs-string.hpp]: Also east const on pointers, yay! west const on other types, awww :(

File: shared/utils/typedefs-string.hpp [hunk 5]
```diff
@@ -83,13 +86,25 @@ struct ConstString {
         return {chars, static_cast<std::size_t>(sz)};
     }
 
+    friend StringW;
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: You can `friend struct StringW` here instead of the FD and the friend.

File: shared/utils/typedefs-string.hpp [hunk 6]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+        return *this;
+    }
+
+    StringW& operator +=(StringW rhs) noexcept {
+        inst = il2cpp_utils::detail::alloc_str(this->operator std::u16string().append(rhs));
+        return *this;
+    }
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Ditto previous

File: shared/utils/typedefs-string.hpp [hunk 7]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+        return strappend(inst, rhs);
+    }
+
+    StringW operator +(StringW rhs) const noexcept {
+        return this->operator std::u16string().append(rhs);
+    }
+
+    bool operator <(StringW rhs) const noexcept;
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Ditto previous

File: shared/utils/typedefs-string.hpp [hunk 8]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+    bool operator <(const T rhs) const noexcept {
+        return strless(inst, rhs);
+    }
+
+    template<int sz>
+    constexpr bool operator ==(ConstString<sz> const& rhs) const noexcept {
+        Il2CppChar* first = inst->chars;
+        char16_t const* second = rhs.chars;
+        Il2CppChar* firstEnd = first + inst->length;
+        char16_t const* secondEnd = second + sz;
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Since it seems `ConstString<sz>` is still constructed from a char array where: `sz` is deduced to be the array length, it means that: `ConstString x = "abc";` will actually have `sz = 4`. This means that in this case, where we start at the beginning of the characters and walk forward, would be incorrect, since if we have a string of `abc` in `this`, and are comparing it to the previous `ConstString`, we will count one extra.

This can be fixed in a few ways, nearly all seem to be reasonable:
1. Make the `sz` for `ConstString` actually be 1 less than the string constant length
2. Use `sz - 1` here, but also note that you would have to make sure that everywhere else where `sz` is used needs to also be aware of that.
3. Use `rhs.length`

Hopefully that gives you some ideas.

File: shared/utils/typedefs-string.hpp [hunk 9]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+            first++; second++;
+        }
+
+        return first == firstEnd && second == secondEnd;
+    }
+    
+    bool operator ==(StringW rhs) const noexcept;
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Ditto previous operators

File: shared/utils/typedefs-string.hpp [hunk 10]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+    bool operator ==(const T rhs) const noexcept {
+        return streq(inst, rhs);
+    }
+
+    bool starts_with(StringW rhs) const noexcept;
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Ditto previous operators

File: shared/utils/typedefs-string.hpp [hunk 11]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+    bool starts_with(const T rhs) const noexcept {
+        return strstart(inst, rhs);
+    }
+    
+    bool ends_with(StringW rhs) const noexcept;
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Ditto previous operators

File: shared/utils/typedefs-string.hpp [hunk 12]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+    bool ends_with(const T rhs) const noexcept {
+        return strend(inst, rhs);
+    }
+    
+    // who needs bounds checking amiright
+    constexpr const Il2CppChar& operator [](size_t idx) const { return inst->chars[idx]; }
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: Very reasonable, though easy const! and also, it might make sense to actually expose a non-const indexing operator for cool people to modify strings in place without having to make new strings and copy, etc. Seems like a simple-enough add, so I'd recommend it.

File: shared/utils/typedefs-string.hpp [hunk 13]
```diff
@@ -102,19 +117,110 @@ struct StringW {
...
+    operator std::string() const;
+    operator std::u16string() const;
+    operator std::wstring() const;
     operator std::u16string_view();
+    operator const std::u16string_view() const;
```
Comment (sc2ad) [inline shared/utils/typedefs-string.hpp]: You guessed it

File: src/utils/typedefs-wrapper.cpp [hunk 14]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+StringW::operator std::u16string() const {
     return {inst->chars, inst->chars + inst->length};
 }
 
-StringW::operator std::wstring() {
+StringW::operator std::wstring() const {
     return {inst->chars, inst->chars + inst->length};
 }
 
+StringW::operator const std::u16string_view() const {
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: Moment

File: src/utils/typedefs-wrapper.cpp [hunk 15]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+    return {inst->chars, static_cast<std::size_t>(inst->length)};
+}
+
 StringW::operator std::u16string_view() {
     return {inst->chars, static_cast<std::size_t>(inst->length)};
 }
+
+bool strcomp(Il2CppString const* lhs, const std::u16string_view rhs) noexcept {
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: See previous statements about const and location of these functions

File: src/utils/typedefs-wrapper.cpp [hunk 16]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+    if (!lhs || lhs->length != (int)rhs.size()) return false;
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: Is this `int` cast necessary? I would hope not, but I don't remember max length of a string, perhaps it makes more sense to cast to `decltype(lhs->length)`? not sure though

File: src/utils/typedefs-wrapper.cpp [hunk 17]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+        }
+        return *first < *second;
+    }
+    // if we got here, and second is not second end, we had a shorter first, so it should be true
+    // if second is the end, we are longer, so it should be false
+    return second != secondEnd;
+}
+
+StringW strappend(Il2CppString const* lhs, const std::string_view rhs) noexcept {
+    return std::u16string(lhs->chars, lhs->chars + lhs->length).append(std::u16string(rhs.data(), rhs.data() + rhs.size()));
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: Consider using the conversion functions to ensure validity instead of using pointer values that are NOT correct (utf8 chars, but creating a string of utf16 chars. 3 characters would parse into a utf16 string of 1 character, where the character would be half one and half the other)

Comment (sc2ad) [reply src/utils/typedefs-wrapper.cpp]: Also note really a huge fan of the double allocation here anyways, the append should be a known size, so we should be able to only need to allocate space for ONE new Il2CppString* and copy all the chars to there element by element. This also feels like something that would be more performant if it could take advantage of the fact that the lhs could be a `StringW` and not an `Il2CppString`, since if it's an `Il2CppString`, we don't have the power to allocate as easily (though we still could just fine).

Basically, I think this function, and where it is used, could stand to avoid these two excess allocations, as well as incorrect behavior with `rhs` conversion. _Ideally_, this function should make a new Il2CppString* that is of the correct size, and copy each character in, promoting the `char` characters to `char16_t`. Hopefully that makes sense.

File: src/utils/typedefs-wrapper.cpp [hunk 18]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+}
+
+StringW strappend(Il2CppString const* lhs, const std::u16string_view rhs) noexcept {
+    return std::u16string(lhs->chars, lhs->chars + lhs->length).append(rhs);
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: Here, we should be able to allocate an `Il2CppString*` with the known lengths and copy both into the new string. This should prevent the excess allocation that we perform by making an `std::u16string`

File: src/utils/typedefs-wrapper.cpp [hunk 19]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+}
+
+bool strstart(Il2CppString const* lhs, const std::string_view rhs) noexcept {
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: On a related note, it seems like many of these `str...` functions could be templated to accept either an `std::string_view` or an `std::u16string_view`. Perhaps that's worth considering?

File: src/utils/typedefs-wrapper.cpp [hunk 20]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+            continue;
+        }
+        // we got a mismatch! return false;
+        return false;
+    }
+    // if we got through the entire string it was all equal, return true
+    return true;
+}
+
+bool StringW::operator ==(StringW rhs) const noexcept {
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: can this be a `StringW const&`?

File: src/utils/typedefs-wrapper.cpp [hunk 21]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+    while (first != firstEnd && second != secondEnd)
+    {
+        if (*first != *second) return false;
+        first++; second++;
+    }
+
+    return first == firstEnd && second == secondEnd;
+}
+
+bool StringW::operator <(StringW rhs) const noexcept {
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: likewise

File: src/utils/typedefs-wrapper.cpp [hunk 22]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+        }
+        return *first < *second;
+    }
+    // if we got here, and second is not second end, we had a shorter first, so it should be true
+    // if second is the end, we are longer, so it should be false
+    return second != secondEnd;
+}
+
+
+bool StringW::starts_with(StringW rhs) const noexcept {
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: likewise

File: src/utils/typedefs-wrapper.cpp [hunk 23]
```diff
@@ -44,26 +46,286 @@ namespace detail {
...
+        }
+        // we got a mismatch! return false;
+        return false;
+    }
+    // if we got through the entire string it was all equal, return true
+    return true;
+}
+
+
+bool StringW::ends_with(StringW rhs) const noexcept {
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: likewise

**Review by sc2ad — CHANGES_REQUESTED** [review]
Oops I wrote like a bunch of comments

File: shared/utils/typedefs-string.hpp (hunk 1 above)
Comment (RedBrumbler) [reply shared/utils/typedefs-string.hpp]: Sorry what do you mean with West const on other types?

File: shared/utils/typedefs-string.hpp [hunk 24]
```diff
@@ -22,11 +22,103 @@ namespace detail {
...
+    }
+        
+    Il2CppString* strappend(Il2CppString const* lhs, Il2CppString const* rhs) noexcept;
+    Il2CppString* strappend(Il2CppString const* lhs, std::u16string_view const rhs) noexcept;
+    Il2CppString* strappend(Il2CppString const* lhs, std::string_view const rhs) noexcept;
+    
+    template<typename T>
+    requires(std::is_same_v<T, std::string_view> || std::is_same_v<T, std::u16string_view>)
+    bool strstart(Il2CppString const* lhs, T const rhs) noexcept {
+        if (!lhs || lhs->length < (decltype(lhs->length))rhs.size()) return false;
```
Comment (RedBrumbler) [inline shared/utils/typedefs-string.hpp]: the cast is needed, but I did change the way it casts to what you suggested

File: shared/utils/typedefs-string.hpp [hunk 25]
```diff
@@ -22,11 +22,103 @@ namespace detail {
...
+            return false;
+        }
+        // if we got through the entire string it was all equal, return true
+        return true;
+    }
+    
+    template<typename T>
+    requires(std::is_same_v<T, std::string_view> || std::is_same_v<T, std::u16string_view>)
+    bool strend(Il2CppString const* lhs, T const rhs) noexcept {
+        if (!lhs || lhs->length < (decltype(lhs->length))rhs.size()) return false;
```
Comment (RedBrumbler) [inline shared/utils/typedefs-string.hpp]: same casting thing

File: src/utils/typedefs-wrapper.cpp [hunk 26]
```diff
@@ -44,26 +46,165 @@ namespace detail {
     }
     Il2CppString* alloc_str(std::u16string_view str) {
         il2cpp_functions::Init();
-        return il2cpp_functions::string_new_utf16((Il2CppChar*) str.data(), str.size());
+        return il2cpp_functions::string_new_utf16((Il2CppChar const*) str.data(), str.size());
+    }
+
+    Il2CppString* strappend(Il2CppString const* lhs, Il2CppString const* rhs) noexcept {
+        if (!lhs && !rhs) return il2cpp_functions::string_new_utf16(u"", 0);
```
Comment (sc2ad) [inline src/utils/typedefs-wrapper.cpp]: do you think it would make more sense to just ret null here? I think it might. Curious what you think though.

File: shared/utils/typedefs-string.hpp (hunk 3 above)
Comment (sc2ad) [reply shared/utils/typedefs-string.hpp]: Adding on this, it should be possible to just say `inst = strappend(inst, rhs)` etc. instead of making a StringW and then casting back to an Il2CppString*

File: src/utils/typedefs-wrapper.cpp (hunk 26 above)
Comment (RedBrumbler) [reply src/utils/typedefs-wrapper.cpp]: hmmm, my thought here was that 2 nullptr strings would be an empty string if appended, but nullptr makes sense too

**Comment by RedBrumbler** [comment]
didn't actually remove it, just removed a typedef

**Review by sc2ad — APPROVED** [review]
